proposals/mdm-baseline.md

```diff
@@ -0,0 +1,153 @@
+  in the application logic exist and the right indexes are created for fast
+  access.
+
+Currently, the code thinks of these two tables as two disjoint sets of elements
+that represent the same entity: a host.
+
+Motivation:
+
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to return hosts
+  that are scheduled for DEP enrollment (not even enrolled yet) to the `GET
+  /hosts` response.
+- In https://github.com/fleetdm/fleet/issues/8878 we're asked to retrieve a
+  list of hosts enrolled via `osquery` that are pending to be enrolled via MDM.
+
+#### Proposed solution
+
+`hosts` should always be the source of truth, we could consider
+`nano_enrollments` the same way we treat  the "auxiliary" `hosts_*` tables (eg:
+`host_disks`): as tables holding metadata for a host. 
+
+When we get information through MDM channels (be it via an enrollment, or via
+the DEP API notifying a new device purchased,) we always get a device serial
+number.
+
+1. Ensure there's a host in the `hosts` table with that serial number (insert
+   one if there isn't.) Try to pre-populate as much info as possible from the
+   MDM payload (eg: `ProductName` is `hosts.hardware_model`)
+1. Create a new entry in the related `nano_*` table pointing to the host.
+
+This implies that we must add an uniqueness constraint for the serial number in both tables.
+
+We should also add a new column or table to track the MDM enrolling status
+("pending," "enrolled") without having to `JOIN` both tables.
+
+#### TODO:
+
+1. Investigate what are the non-optional fields in the `hosts` table, and if
+   having default empty values for some of them requires a breaking change.
+1. Investigate the consequences of using a serial number to uniquely identify
+   hosts, eg: what happens if two users on the same machine use different
+   enrollment profiles? See "macOS extensions" in MDM spec for details.
+
+### Problem: Issue an MDM command and monitor it’s status in fleet
+
+As described in
+[#Commands](./tools/mdm/apple/glossary-and-protocols.md#commands), sending and
+getting results from an MDM command is a multi stage process. 
+
+We should:
+
+1. Ensure all potential errors in each step are handled and _recorded_, so we
+   can report back to the user if there was an error.
+1. Ensure we have endpoints with desirable interfaces to run and get
+   information.
+
+#### Proposed solution
+
+1. Research and understand all the failure points when executing a command.
+1. Research and understand why issues like
+   [these](https://github.com/fleetdm/fleet/issues/8815) happen and figure out
+   if there's anything we can do.
+1. Ensure we're handling all errors as described in the MDM spec, including
+   `NotNow` commands.
+1. All MDM responses are XML, the server will be in charge of transforming those to JSON.
+    1. Discuss: this results in a heavier load in the server, should we
+       consider just delivering XML to the client to "distribute" this effort?
```

Comment:
I find it hard to believe the traffic from MDM APIs to be heavy enough that translating XML to JSON would become an issue. Right? What traffic do we expect from this?

Reply:
+1 we were thinking the same thing, and addressing the performance issue when it becomes a problem

Reply:
which makes me think, we will have to tweak osquery-perf to simulate MDM payloads too, right?

Reply:
Maybe. Not sure if we want osquery-perf to play that part as well... at this point we should call it fleet-perf though hah